File: bluetooth/lib/src/bluetooth.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:my_bluetooth/my_bluetooth_mixin.dart';
import 'package:my_bluetooth/my_bluetooth_util.dart';

class Bluetooth extends StatefulWidget {
  final Control? parent;
  final Control control;
  final FletControlBackend backend;

  const Bluetooth(
      {super.key, this.parent, required this.control, required this.backend});
  @override
  _Bluetooth createState() => _Bluetooth();
}

class _Bluetooth extends State<Bluetooth> with MyBluetoothMixin<Bluetooth>{
  late TextEditingController _controller;
  String _string = "";
  bool _set_channel_name = true;
  // bool _start_scan = false;
  // bool _isConnect = false;
  // late List<String> _bluetooth_list;
  // late int _connect_message = 0;
  MyBluetoothUtil myBluetoothUtil = MyBluetoothUtil();

  @override
  Widget build(BuildContext context) {
    var channel_name = widget.control.attrString("channel_name");
    if (_set_channel_name && channel_name != null) {
      myBluetoothUtil.setMessageChannel(channel_name, listenerAndroidHandle);
      myBluetoothUtil.sendChannelName("channelName", channel_name);
      setState(() {
        _set_channel_name = false;
      });
    }
    var message_tag = widget.control.attrString("message_tag");
    var start_scan = widget.control.attrBool("start_scan")!;
    var stop_scan = widget.control.attrBool("stop_scan")!;
    var isConnect = widget.control.attrBool("isConnect")!;
    var close_connect = widget.control.attrBool("close_connect")!;
    var start_reader = widget.control.attrBool("start_reader")!;
    var start_reader_epc = widget.control.attrBool("start_reader_epc")!;
    var start_set_antenna = widget.control.attrBool("start_set_antenna")!;
    var start_set_antenna_power = widget.control.attrBool("start_set_antenna_power")!;
    var query_rfid_capacity = widget.control.attrBool("query_rfid_capacity")!;
    var name = widget.control.attrString("name");
    var address = widget.control.attrString("address");
    var antenna_num = widget.control.attrInt("antenna_num");
    var antenna_num_power = widget.control.attrString("antenna_num_power");
    var bluetooth_list = widget.control.attrString("bluetooth_list");
    var connect_message = widget.control.attrString("connect_message");
    var scanner_message = widget.control.attrString("scanner_message");
    var reader_operation_message = widget.control.attrString("reader_operation_message");
    var epc_messages = widget.control.attrString("epc_messages");
    var antenna_message = widget.control.attrString("antenna_message");
    var rfid_capacity_message = widget.control.attrString("rfid_capacity_message");
    
    if (start_scan) {     // 开始扫描标志为True时向Android端发送对应信息
      myBluetoothUtil.sendMessageToAndroid("startScanner", true);
      setState(() {
        /*
          发送完消息后将开始扫描标志置为“false”（目前只掌握更新字符串属性方式，
          不过影响不大python控件那边布尔型属性可以接受字符串）
          下方多个标志一样的处理方式
        */
        widget.backend.updateControlState(widget.control.id, {"start_scan": "false"});
      });
    } 
    if (stop_scan)  {
      myBluetoothUtil.sendMessageToAndroid("stopScanner", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"stop_scan": "false"});
      });
    }
    if (start_set_antenna_power && antenna_num_power != "") {
      myBluetoothUtil.sendMessageToAndroid("setAntennaPower", antenna_num_power);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"start_set_antenna_power": "false"});
      });
    }
    if (isConnect && address != "") {    // 连接设备标志
      myBluetoothUtil.sendMessageToAndroid("bluetoothAddress", address);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"isConnect": "false"});
      });
    }
    if (close_connect) {    // 关闭设备连接标志
      myBluetoothUtil.sendMessageToAndroid("closeConnect", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"close_connect": "false"});
      });
    }

    if (start_reader) {
      myBluetoothUtil.sendMessageToAndroid("startReader", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"start_reader": "false"});
      });
    } 
    if (start_reader_epc) {
      myBluetoothUtil.sendMessageToAndroid("startReaderEpc", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"start_reader_epc": "false"});
      });
    }
    if (start_set_antenna && antenna_num != 0) {
      myBluetoothUtil.sendMessageToAndroid("setAntennaNum", antenna_num);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"start_set_antenna": "false"});
      });
    }
    if (query_rfid_capacity) {
      myBluetoothUtil.sendMessageToAndroid("queryRfidCapacity", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"query_rfid_capacity": "false"});
      });
    }
    if (_string != "") {
      _controller.text = _string;
    }
    Widget bluetooth_control = TextField(
        controller: _controller,
        decoration: const InputDecoration(
          border: InputBorder.none, //无边框
        )
    );
    return constrainedControl(
        context, bluetooth_control, widget.parent, widget.control);
  }

  @override
  Future<void> listenerAndroidHandle(dynamic message) async {
    // TODO: implement listenerAndroidHandle
    var scannerMessage = message["scanMessage"];
    if (scannerMessage != null) {
      // 开始扫描信息
      widget.backend.updateControlState(widget.control.id, {"scanner_message": scannerMessage});
      String eventString = scannerMessage;
      widget.backend.updateControlState(widget.control.id, {"message_tag": "scanner_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 设备列表
    var bluetoothList = message["bluetooth_list"];
    if (bluetoothList != null) {
      String eventString = "";
      for (String bluetooth in bluetoothList) {
        eventString += "$bluetooth&";
      }
      widget.backend.updateControlState(widget.control.id, {"bluetooth_list": eventString});
      widget.backend.updateControlState(widget.control.id, {"message_tag": "bluetooth_list"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 连接结果（连接成功、连接失败、断开连接）消息样式为   连接成功>>>（设备名称）
    var connectMessage = message["connectMessage"];
    if (connectMessage != null) {
      widget.backend.updateControlState(widget.control.id, {"connect_message": connectMessage});
      String eventString = connectMessage;
      widget.backend.updateControlState(widget.control.id, {"message_tag": "connect_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 读写器读卡操作信息
    var readerOperationMessage = message["readerOperationMessage"];
    if (readerOperationMessage != null) { 
      widget.backend.updateControlState(widget.control.id, {"reader_operation_message": readerOperationMessage});
      String eventString = readerOperationMessage;
      widget.backend.updateControlState(widget.control.id, {"message_tag": "reader_operation_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 读取epc数据
    var epcMessages = message["epcMessages"];
    if (epcMessages != null) {
      String eventString = "";
      for (String epcMessage in epcMessages)  {
        eventString += "$epcMessage&";
      }
      widget.backend.updateControlState(widget.control.id, {"epc_messages": eventString});
      widget.backend.updateControlState(widget.control.id, {"message_tag": "epc_messages"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 天线配置信息
    var antenna_num_message = message["AntennaNumMessage"];
    if (antenna_num_message != null) {
      String eventString = antenna_num_message;
      widget.backend.updateControlState(widget.control.id, {"antenna_message": eventString});
      widget.backend.updateControlState(widget.control.id, {"message_tag": "antenna_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // RFID读写器读写能力信息
    var rfid_capacity_message = message["rfidCapacityMessage"];
    if (rfid_capacity_message != null) {
      String eventString = rfid_capacity_message;
      widget.backend.updateControlState(widget.control.id, {"rfid_capacity_message": eventString});
      widget.backend.updateControlState(widget.control.id, {"message_tag": "rfid_capacity_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
  }
}

File: python/controls/bluetooth.py
from flet_core.constrained_control import ConstrainedControl
from typing import Optional

class Bluetooth(ConstrainedControl):
    """
    Bluetooth 控件
    """

    def __init__(
        self,
        name: Optional[str] = "",
        address: Optional[str] = "",
        antenna_num: Optional[str] = 0,
        antenna_num_power: Optional[str] = "",
        message_tag: Optional[str] = "",
        start_scan: Optional[bool] = False,
        stop_scan: Optional[bool] = False,
        isConnect: Optional[bool] = False,
        close_connect: Optional[bool] = False,
        start_reader: Optional[bool] = False,
        start_reader_epc: Optional[bool] = False,
        start_set_antenna: Optional[bool] = False,
        start_set_antenna_power: Optional[bool] = False,
        query_rfid_capacity: Optional[bool] = False,
        bluetooth_list: Optional[str] = "",
        connect_message: Optional[str] = "",
        scanner_message: Optional[str] = "",
        epc_messages: Optional[str] = "",
        reader_operation_message: Optional[str] = "",
        antenna_message: Optional[str] = "",
        rfid_capacity_message: Optional[str] = "",
        on_listener = None,
    ):
        ConstrainedControl.__init__(self)
        self.name = name         # 设备名称（选定设备时设置）
        self.address = address   # 设备mac地址 （选定时设置）
        self.antenna_num = antenna_num   #  设备天线端口
        self.antenna_num_power = antenna_num_power # 设置天线端口功率
        self.start_scan = start_scan    # 开始扫描设备标志
        self.stop_scan = stop_scan     # 停止扫描设备标志
        self.start_reader = start_reader   # 启动读写器读卡标志
        self.start_reader_epc = start_reader_epc    # 读取epc标签数据标志
        self.start_set_antenna = start_set_antenna  # 设置读卡天线端口标志
        self.start_set_antenna_power = start_set_antenna_power  # 设置天线端口功率标志
        self.isConnect = isConnect          # 连接设备标志
        self.bluetooth_list = bluetooth_list     # 扫描到的设备列表
        self.connect_message = connect_message    # 连接设备时返回的提示信息
        self.scanner_message = scanner_message     # 开始扫描时返回的提示信息
        self.epc_messages = epc_messages    # 单次盘点所获取的epc标签数据
        self.on_listener = on_listener     # 用于监听Android端发送过来的信息事件方法
        self.message_tag = message_tag    # 用于甄别Android端发送过来的信息是何种类型
        self.close_connect = close_connect     # 断开设备连接标志
        self.reader_operation_message = reader_operation_message    # 启动读写器读卡返回的信息
        self.antenna_message = antenna_message    # 设置读卡天线端口信息
        self.query_rfid_capacity = query_rfid_capacity    #  查询rfid读写器的读写能力信息
        self.rfid_capacity_message = rfid_capacity_message     #  rfid读写器的读写能力信息
        self.channel_name = str(id(self))


    def _get_control_name(self) -> str:
        return "bluetooth"


    @property
    def name(self):
        return self._get_attr("name")

    @name.setter
    def name(self, value):
        self._set_attr("name", value)


    @property
    def address(self):
        return self._get_attr("address")

    @address.setter
    def address(self, value):
        self._set_attr("address", value)

    @property
    def start_scan(self):
        return self._get_attr("start_scan")

    @start_scan.setter
    def start_scan(self, value):
        self._set_attr("start_scan", value)

    @property
    def stop_scan(self):
        return self._get_attr("stop_scan")

    @stop_scan.setter
    def stop_scan(self, value):
        self._set_attr("stop_scan", value)

    @property
    def isConnect(self):
        return self._get_attr("isConnect")

    @isConnect.setter
    def isConnect(self, value):
        self._set_attr("isConnect", value)

    @property
    def close_connect(self):
        return self._get_attr("close_connect")

    @close_connect.setter
    def close_connect(self, value):
        self._set_attr("close_connect", value)

    @property
    def start_reader(self):
        return self._get_attr("start_reader")

    @start_reader.setter
    def start_reader(self, value):
        self._set_attr("start_reader", value)

    @property
    def start_reader_epc(self):
        return self._get_attr("start_reader_epc")

    @start_reader_epc.setter
    def start_reader_epc(self, value):
        self._set_attr("start_reader_epc", value)

    @property
    def bluetooth_list(self):
        return self._get_attr("bluetooth_list")

    @bluetooth_list.setter
    def bluetooth_list(self, value):
        self._set_attr("bluetooth_list", value)

    @property
    def connect_message(self):
        return self._get_attr("connect_message")

    @connect_message.setter
    def connect_message(self, value):
        self._set_attr("connect_message", value)

    @property
    def scanner_message(self):
        return self._get_attr("scanner_message")

    @scanner_message.setter
    def scanner_message(self, value):
        self._set_attr("scanner_message", value)

    @property
    def epc_messages(self):
        return self._get_attr("epc_messages")

    @epc_messages.setter
    def epc_messages(self, value):
        self._set_attr("epc_messages", value)

    @property
    def on_listener(self):
        return self._get_event_handler("on_listener")

    @on_listener.setter
    def on_listener(self, handler):
        self._add_event_handler("on_listener", handler)
        # self._set_attr("start_listener", True if handler is not None else None)
    @property
    def message_tag(self):
        return self._get_attr("message_tag")

    @message_tag.setter
    def message_tag(self, value):
        self._set_attr("message_tag", value)

    @property
    def reader_operation_message(self):
        return self._get_attr("reader_operation_message")

    @reader_operation_message.setter
    def reader_operation_message(self, value):
        self._set_attr("reader_operation_message", value)

    @stop_scan.setter
    def stop_scan(self, value):
        self._set_attr("stop_scan", value)

    @property
    def antenna_num(self):
        return self._get_attr("antenna_num")

    @antenna_num.setter
    def antenna_num(self, value):
        self._set_attr("antenna_num", value)

    @property
    def start_set_antenna(self):
        return self._get_attr("start_set_antenna")

    @start_set_antenna.setter
    def start_set_antenna(self, value):
        self._set_attr("start_set_antenna", value)

    @property
    def antenna_message(self):
        return self._get_attr("antenna_message")

    @antenna_message.setter
    def antenna_message(self, value):
        self._set_attr("antenna_message", value)

    @property
    def antenna_num_power(self):
        return self._get_attr("antenna_num_power")

    @antenna_num_power.setter
    def antenna_num_power(self, value):
        self._set_attr("antenna_num_power", value)

    @property
    def start_set_antenna_power(self):
        return self._get_attr("start_set_antenna_power")

    @start_set_antenna_power.setter
    def start_set_antenna_power(self, value):
        self._set_attr("start_set_antenna_power", value)

    @property
    def channel_name(self):
        return self._get_attr("channel_name")

    @channel_name.setter
    def channel_name(self, value):
        self._set_attr("channel_name", value)
    
    @property
    def query_rfid_capacity(self):
        return self._get_attr("query_rfid_capacity")

    @query_rfid_capacity.setter
    def query_rfid_capacity(self, value):
        self._set_attr("query_rfid_capacity", value)
    
    @property
    def rfid_capacity_message(self):
        return self._get_attr("rfid_capacity_message")

    @rfid_capacity_message.setter
    def rfid_capacity_message(self, value):
        self._set_attr("rfid_capacity_message", value)
    
    def start_scanner(self):
        self._set_attr("start_scan", True)
    
    def stop_scanner(self):
        self._set_attr("stop_scan", True)
    
    def connect(self, bluetooth_address):
        self._set_attr("address", bluetooth_address)
        self._set_attr("isConnect", True)
    
    def close(self):
        self._set_attr("close_connect", True)
    
    def reader(self):
        self._set_attr("start_reader", True)
    
    def reader_epc(self):
        self._set_attr("start_reader_epc", True)
    
    def set_antenna_num(self, antenna_num):
        self._set_attr("antenna_num", antenna_num)
        self._set_attr("start_set_antenna", True)
    
    def set_antenna_power(self, antenna_powers):
        antenna_power_str = ""
        for antenna_power in antenna_powers:
            antenna_num = antenna_power["antenna"]
            power = antenna_power["power"]
            antenna_power_str += f"{antenna_num}#{power}&"
        self._set_attr("antenna_num_power", antenna_power_str[:-1])
        self._set_attr("start_set_antenna_power", True)
    
    def query(self):
        self._set_attr("query_rfid_capacity", True)




